### ISSUE TITLE
Inference on unseen data without ground truth pose and transformation information

### ISSUE THREAD

**Issue by gowda-95** [hxwork/OMNet_Pytorch#7]
Hello,

thank you for the nice work! I have trained the model on a custom dataset. But i need to perform inference on unseen data without ground truth pose and transformation values.  The test(model, manager) function in evaluate.py requires "transform_gt" and "pose_gt" to generate_overlap_mask at line 37 (output_batch = model(data_batch).

Could you please provide codes for inference on unseen pointclouds (without ground truths) or let me know how it can be done. 

Thank you.

**Comment by hxwork** [comment]
Hi,

Thanks for your interest. I'm sorry that I have no inference code right now, but I think it's easy to achieve your goal by only modifying our code slightly. For example, you can just remove everything which is related to the ground truth of the pose and transformation since we don't use them for inference. More specifically, you can remove the lines in [net.py](https://github.com/hxwork/OMNet_Pytorch/blob/main/model/net.py) which are related to [L40](https://github.com/hxwork/OMNet_Pytorch/blob/75dac389cae48e2d6579752a1dd7d3a8e68956e1/model/net.py#L40) and [L41](https://github.com/hxwork/OMNet_Pytorch/blob/75dac389cae48e2d6579752a1dd7d3a8e68956e1/model/net.py#L41).

Regards,
Hao Xu

**Comment by gowda-95** [comment]
Hi,

Many thanks for your quick reply. I made the suggested changes in [net.py](https://github.com/hxwork/OMNet_Pytorch/blob/main/model/net.py#L104) 
![image](https://user-images.githubusercontent.com/103493119/196372709-4821c7d3-c9bb-417a-9e04-84231caf0b48.png)
as shown in the image. And I am trying to save the outputs as by modifying [evaluate.py](https://github.com/hxwork/OMNet_Pytorch/blob/main/evaluate.py#L66). 
I am getting a list of dictionary of arrays as the output. 
[output.zip](https://github.com/hxwork/OMNet_Pytorch/files/9808267/output.zip)

My initial assumption was ["all_src_cls_pair"](https://github.com/hxwork/OMNet_Pytorch/blob/main/model/net.py#L109) was the transformed source point cloud, but the shape( (['all_src_cls_pair'][0][0]).shape) I am getting is (1,2,2048). 

Could you please let me know if my understanding that 'all_src_cls_pair' was transformed source point cloud was wrong, and if so how do I get inference of transformed source point cloud.

Thanks again. Best regards,
Nishanth

**Comment by hxwork** [comment]
Hi,

Note that `all_src_cls_pair` and `all_ref_cls_pair` are the predicted classification results and the corresponding classification ground truth at each iteration of the source and reference point clouds respectively. 

For the transformed source point cloud, please use `transform_pred` or `pose_pred` output from the last iteration to transform the input source point cloud.

Best,
Hao

**Comment by gowda-95** [comment]
Hi,

Thanks again for the quick reply.

**Comment by gowda-95** [comment]
Hi,

I had another doubt. I did run inference and saved the [transform_pred](https://github.com/hxwork/OMNet_Pytorch/blob/main/model/net.py#L113). I then used the function [apply_transform](https://github.com/hxwork/OMNet_Pytorch/blob/main/dataset/transformations.py#L299) and the saved [transform_pred](https://github.com/hxwork/OMNet_Pytorch/blob/main/model/net.py#L113). 

But I am getting a large translation to various inputs, so i was hoping to verify if [apply_transform](https://github.com/hxwork/OMNet_Pytorch/blob/main/dataset/transformations.py#L299) is the right function to use. 

Thanks again and Best regards,
Nishanth

**Comment by hxwork** [comment]
Hi,

If the saved result is a 3D transformation matrix, which contains a $3\times 3$ rotation matrix and a $3\times 1$ translation vector, then you can use the `apply_transform` function to perform the transformation. Note that you should make sure the direction (src -> ref or ref -> src) is correct.

Best,
Hao

**Comment by gowda-95** [comment]
Hi, 

Many thanks for the quick reply. I think I have figured out the error. The pointclouds in ModelNet40 dataset used by you are 0 centred. But the ground truth I generated to train the model the point clouds are not zero centred. Is it necessary for the point clouds to be zero centred while training the network? Or just changing the apply transform to account for the present centre of the point cloud should be okay?

Thank you.

Best regards,
Nishanth
